### DISCUSSION TITLE
[Question] Units and measurements of entities

### DISCUSSION THREAD

**Discussion by typhoon71** [mozman/ezdxf#818] [Q&A]
Sorry for the noobish question, but I couldn't find a way to solve this myself.

I'm creating a drawing from scratch, and ezdxf gives me almost the expected result.
I use something like this:

```python3
doc = ezdxf.new('R2007', setup=True)  # I will be using R12 in the end, but I tried with all of them
doc.units = units.MM

msp = doc.modelspace()

doc.layers.add(name='mylayer', color=3, linetype='CONTINUOS')
msp.add_line([2,3], [2,5])  # random example

zoom.extents(msp)  # edit: fixed this line
doc.saveas('test.dxf')
```

My problem is that if I I draw a line from point [2,3] to point [2,5] and measure it in any CAD application, it's not 2 as I'd expect, but 200!
I'm "drawing" in millimeters, so units.MM should've set 1 unit to 1mm... or I'm wrong?

Can someone point me to the right direction? Thanks a lot.

**Comment by mozman** [comment]
The Measure tool in Autodesk Trueview measures 2.0000:

![image](https://user-images.githubusercontent.com/4230966/211027667-4f5ba61a-d634-4368-94fb-41d3153f34dd.png)

What is that "any" CAD application which measures 200?

**Reply by typhoon71** [reply]
Yeah, sorry, "any" was not really informative.
Right now I can test it only with Librecad/Nanocad/Intellicad - TLDR: no access to good CAD SW right now.

FYI: if I examine the entity properties lenghts is correct (2), and if I paste my drawing on another "correctly working" one the scale is perfect.

In the end I found an option called "LENGHT FACTOR" or "MEASUREMENT SCALE" which value was 100... suspicious... I changed it to 1 and now my measurement are what I expect them to be (that line measures 2).

The thing that puzzles me is that I didn't create any of such setting! Can it be that it's set to 100 by `setup=True`?  Or maybe if a measurement scale is not set those cad sw I used go bonkers? 

Next week I'll be able to test if this happens with Autocad or Bricscad - I'll let you know.

EDIT: Autodesk Trueview 2019 cannot open the .dxf generated from the code in my first post - how did you manage to? I'm missing something here.

**Comment by typhoon71** [comment]
The issue I'm experiencing seems to be caused by $DIMLFAC being set to 100 instead of 1.

If you open the dxf generated by the code sample in my first post you'll notice the $DIMLFAC block is set to 100.0. 

If I change it with 1.0 (editing the dxf file with a text editor) I get the meaurements I'm expecting.

I'm using ezdxf 1.0.0, python 3.8

**Comment by typhoon71** [comment]
I examined ezdxf code a bit.

If I read standards.py correctly, when creating a new dxf document with `setup=True` the default dimension style always has `$DIMLFAC=100`, because `EZ_M_100_H25_CM` is used (always take `M` and `CM`).

Creating the dxf document with `doc = ezdxf.new('R2007', setup=True, units=units.MM)` doesn't fix the issue (the `units` switch is ignored).

I'd expect it to be honored - if passed at document creation.

**Comment by typhoon71** [comment]
The same happens with BricsCAD v23, all measure are 100x.
This is on a modern workstation, where BricsCAD opens the .dxf while Autodesk Trueview fails; puzzling.

I can work around the issue with `doc.header['$DIMLFAC'] = 1`.

I think it would be cleaner if `doc = ezdxf.new('R2007', setup=True, units=units.MM) ` would actually be honored, i.e. being used in the dimension style creation.

Do you think I should open a separate issue as a bug?

**Comment by mozman** [comment]
The settings in the HEADER section are evaluated only by CAD applications, and not by all CAD applications to the same extend. The `ezdxf` package is a tool to create DXF documents and not to control CAD applications, the decision of Autodesk to store the current application settings in the DXF file was a bad idea for a data exchange format (e.g. JPG files do not contain Photoshop settings).

If you use mm as drawing units, does not mean you always want dimensions in mm, that is just your current expectation, which may not match all users needs. So setting `$DIMFLAC` to 1 is not a workaround, it's something you have to do to get the desired result or setup your own DIMSTYLE in the CAD application or by [ezdxf](https://ezdxf.mozman.at/docs/tables/dimstyle_table_entry.html) and set it as default DIMSTYLE.

There may be other settings stored in the 200+ header variables that do not meet your needs.

> This is on a modern workstation, where BricsCAD opens the .dxf while Autodesk Trueview fails; puzzling.

When Trueview can't open your DXF document, it's not related to the `$DIMFLAC` header variable.

EDIT: 
When you created your own DIMSTYLE, there exist the method [DimStyle.copy_to_header()](https://ezdxf.mozman.at/docs/tables/dimstyle_table_entry.html#ezdxf.entities.DimStyle.copy_to_header) to copy all DIMSTYLE attributes starting with "dim" to the HEADER section of a DXF document. That also sets this DIMSTYLE as the default dimension style in the CAD application.

**Comment by typhoon71** [comment]
Thanks for your answer; no I get your point about the header settings, it was a bit confusing for me before.

1) In my case I don't really need to have measurement in the drawing, I just needed to check some key distances in the dwg while I'm developing. I don't even need to setup a proper DIMSTYLE (at least for the foreseeable future).

In the end I created dxf with just:

```python
doc = ezdxf.new('R2007)
doc.units = units.MM

msp = doc.modelspace()

doc.layers.add(name='mylayer', color=3, linetype='CONTINUOS')
msp.add_line([2,3], [2,5])  # random example

doc.saveas('test.dxf')
```

This lets me gets the results I want (measusements are shown in mm) - without messing with header settings.

2) I still have a question: do you have any idea why Trueview can't open your DXF document?
I have no issues opening a dxf created with other applicationas (NanoCAD, Progesoft, BricsCAD, even LibreCAD).

I won't probably be opening my dxf documents with an Autodesk application, but I'd  like to know if I can do something to make them... more likeable.

Could you help me understanding what is happening? If it's not something that would take too much of your time I'd like some help.

**Comment by mozman** [comment]
I don't really know why TrueView does not open your DXF files, because TrueView opens all DXF file I create with `ezdxf`.

But I guess you use a line type, text style or dimension style without an entry in the according resource table. If you want to be sure that your DXF file is valid, call the `doc.validate(print_report=True)` method to check for errors.

**Comment by typhoon71** [comment]
I'll need to investigate this: the sample above is how I generate my dxf documents (just with more lines/arcs).

`doc.validate(print_report=True)` returns True and shows nothing...

The only thing I set is the linetype and color... so it should be the part I'm doing wrong... will see if I catch it and let you know.

**Reply by mozman** [reply]
Hidden in plain sight: "CONTINUOUS" != "CONTINUOS"

The validate method fixes this issue by removing the invalid linetype name and the default linetype is BYLAYER which is CONTINUOUS for layers without a layer table entry.

You can verify this by: `doc.audit().print_fixed_errors()`

**Comment by typhoon71** [comment]
Indeed fixing that typo was the solution! Completely missed it.

Now I can open my dxf in Trueview too. Thanks a lot for your help.
